File: 876.链表的中间节点/middleNode.go
package problem0876

type ListNode struct {
	Val  int
	Next *ListNode
}

// 需要找到链表的中间节点，偶数个时返回第二个
// 1. 遍历链表，放进数组，返回对应索引的节点
// 2. 遍历一遍，获得长度，再遍历即可
// 3. 快慢指针，fast一次走两步，slow一次走一步，当fast为nil时，slow必然再中间
func middleNode(head *ListNode) *ListNode {
	length, headBackup := 0, head
	for head != nil {
		length++
		head = head.Next
	}
	mid := (length >> 1) + 1
	for headBackup != nil {
		mid--
		if mid == 0 {
			return headBackup
		}
		headBackup = headBackup.Next
	}
	return head
}

func middleNode1(head *ListNode) *ListNode {
	slow, fast := head, head
	for fast != nil && fast.Next != nil {
		slow = slow.Next
		fast = fast.Next.Next
	}
	return slow
}

File: 876.链表的中间节点/middleNode.js
// 链表的中间节点
// 给定一个头结点为 head 的非空单链表，返回链表的中间结点。如果有两个中间结点，则返回第二个中间结点。

function middleNode(head) {
    let slow = head
    let fast = head
    // 快指针能够前进的条件：当前节点和下一个节点都为空
    while (fast !== null && fast.next !== null) {
        slow = slow.next
        fast = fast.next.next
    }
    return slow
}

// 如果有两个中间节点时，返回第一个中间节点的解决办法：在链表头添加一个dummy节点
function middleNode2(head) {
    // or new ListNode()
    const dummy = { next: head }
    let slow = dummy
    let fast = dummy
    while (fast !== null && fast.next !== null) {
        slow = slow.next
        fast = fast.next.next
    }
    return slow
}